pkg/submitter/submitter.go

```diff
@@ -0,0 +1,494 @@
+		case <-s.ctx.Done():
+			return nil
+		case result := <-s.resultCh:
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			var ctx context.Context
+			ctx, s.lastSubmitCncl = context.WithCancel(s.ctx)
+			s.handleSubmit(ctx, result)
+		}
+	}
+}
+
+func (s *Submitter) Stop() {
+	if s.lastSubmitCncl != nil {
+		s.lastSubmitCncl()
+	}
+	s.close()
+	level.Info(s.logger).Log("msg", "submitter shutdown complete")
+}
+
+func (s *Submitter) handleSubmit(ctx context.Context, result *mining.Result) {
+	go func(ctx context.Context, result *mining.Result) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(15 * time.Second):
+				level.Debug(s.logger).Log("msg", "retrying submit after 15 seconds")
+				var lastSubmit time.Duration
+				var err error
+				lastSubmit, err = s.lastSubmit()
+				if err != nil {
+					level.Debug(s.logger).Log("msg", "checking last submit time", "err", err)
+					continue
+				}
+				// Initialize.
+				timeToSubmit, cncl := context.WithTimeout(ctx, 0)
+				if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
+					level.Debug(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
+					timeToSubmit, cncl = context.WithTimeout(ctx, s.cfg.Mine.MinSubmitPeriod.Duration-lastSubmit+10*time.Second) // Adding the 10 seconds penalty to be sure we send the tx in the right time!
+				}
+				defer cncl()
+				select {
+				case <-ctx.Done(): // The context was canceled from the Start loop because new work arrived.
+					return
+				case <-timeToSubmit.Done(): // It is time to submit so start the submit loop
+
+					// Set this solution as pending so that if
+					// any of the checks below fail and will be retried
+					// when there is no new challenge.
+					s.solutionPending = result
+
+					profitPercent, err := s.profit() // Call it regardless of whether we use so that is sets the exposed metrics.
+					if err != nil {
+						level.Error(s.logger).Log("msg", "submit solution profit check", "err", err)
+					}
+					if s.cfg.Mine.ProfitThreshold > 0 {
+						if profitPercent < int64(s.cfg.Mine.ProfitThreshold) {
+							level.Debug(s.logger).Log("msg", "transaction not profitable, so will wait for the next cycle")
+						} else { // transaction is profitable.
+							tx, err := s.Submit(s.ctx, result)
+							if err != nil {
+								s.submitFailCount.Inc()
+								level.Error(s.logger).Log("msg", "submiting a solution, retrying", "err", err, "account", s.account.Address.String())
+								continue
+							}
+							level.Debug(s.logger).Log("msg", "submited a solution", "txHash", tx.Hash().String(), "account", s.account.Address.String())
+							s.saveGasUsed(ctx, tx)
+							s.submitCount.Inc()
+							// A solution has been submitted so the
+							// pending solution doesn't matter here any more so reset it.
+							s.solutionPending = nil
+							return
+						}
+					}
+				}
+
+			}
+		}
```

Comment:
This needs a bit of refactoring.
Here is what i have in mind.

```suggestion
func (s *Submitter) blockUntilTimeToSubmit(ctx context.Context) {
	var (
		lastSubmit time.Duration
		timestamp  *time.Time
		err        error
	)
	for {
		select {
		case <-ctx.Done(): // The context was canceled from the main loop because new work arrived.
		default:
		}
		lastSubmit, timestamp, err = s.lastSubmit()
		if err != nil {
			level.Debug(s.logger).Log("msg", "checking last submit time", "err", err)
			time.Sleep(1 * time.Second)
			continue
		}
		break
	}
	var (
		timeToSubmit context.Context
		cncl        context.CancelFunc
	)
	if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
		level.Debug(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
		timeToSubmit, cncl = context.WithDeadline(ctx, timestamp.Add(15*time.Minute))
		defer cncl()
	}

	select {
	case <-ctx.Done(): // The context was canceled from the main loop because new work arrived.
	case <-timeToSubmit.Done(): // 15min since last submit has passed to can unblock.
	}
}

func (s *Submitter) handleSubmit(ctx context.Context, result *mining.Result) {
	go func(ctx context.Context, result *mining.Result) {
		ticker := time.NewTicker(15 * time.Second)
		for {
			select {
			case <-ctx.Done(): // The context was canceled from the main loop because new work arrived.
				return
			default:
				profitPercent, err := s.profit() // Call it regardless of whether we use so that is sets the exposed metrics.
				if err != nil {
					level.Error(s.logger).Log("msg", "submit solution profit check", "err", err)
					<-ticker.C
					continue
				}
				if s.cfg.Mine.ProfitThreshold > 0 {
					if profitPercent < int64(s.cfg.Mine.ProfitThreshold) {
						level.Debug(s.logger).Log("msg", "transaction not profitable, so will wait for the next cycle")
						<-ticker.C
						continue
					} else { // Transaction is profitable.
						if statusID = s.getMinerStatus(); statusID!=... { // I think status ID 3 was the status that allows to submit, but not sure need to double check the contract.
							level.Error(s.logger).Log("msg","miner is not in a status that can submit","statusID",statusID)
							<-ticker.C
							continue
						}
						tx, err := s.Submit(s.ctx, result)
						if err != nil {
							s.submitFailCount.Inc()
							level.Error(s.logger).Log("msg", "submiting a solution, retrying", "err", err, "account", s.account.Address.String())
							<-ticker.C
							continue
						}
						level.Debug(s.logger).Log("msg", "submited a solution", "txHash", tx.Hash().String(), "account", s.account.Address.String())
						s.saveGasUsed(ctx, tx)
						s.submitCount.Inc()
						return
					}
				}

			}
		}
	}(ctx, result)
}
```